reccmp/isledecomp/compare/db.py

```diff
@@ -388,52 +387,79 @@ def is_vtordisp(self, recomp_addr: int) -> bool:
+    def search_name(self, name: str, compare_type: SymbolType) -> Iterator[MatchInfo]:
+        if "name" not in self._indexed:
+            self._sql.execute(
+                "CREATE index idx_name on symbols(json_extract(kvstore, '$.name'))"
+            )
+            self._indexed.add("name")
 
-        return None
+        # n.b. If the name matches and the type is not set, we will return the row.
+        # Ideally we would have perfect information on the recomp side and not need to do this
+        cur = self._sql.execute(
+            """SELECT orig_addr, recomp_addr, kvstore FROM symbols
+            WHERE json_extract(kvstore, '$.name') = ?
+            AND (json_extract(kvstore, '$.type') IS NULL OR json_extract(kvstore, '$.type') = ?)""",
+            (name, compare_type),
+        )
+        cur.row_factory = matchinfo_factory
+        yield from cur
 
     def _match_on(self, compare_type: SymbolType, addr: int, name: str) -> bool:
-        # Update the compare_type here too since the marker tells us what we should do
+        """Search the program listing for the given name and type, then assign the
+        given address to the first unmatched result."""
+        # If we identify the name as a linker symbol, search for that instead.
+        # TODO: Will need a customizable "name_is_symbol" function for other platforms
+        if compare_type != SymbolType.STRING and name.startswith("?"):
+            for obj in self.search_symbol(name):
+                if obj.orig_addr is None and obj.recomp_addr is not None:
+                    return self.set_pair(addr, obj.recomp_addr, compare_type)
+
+            return False
 
         # Truncate the name to 255 characters. It will not be possible to match a name
-        # longer than that because MSVC truncates the debug symbols to this length.
+        # longer than that because MSVC truncates to this length.
         # See also: warning C4786.
         name = name[:255]
 
-        logger.debug("Looking for %s %s", compare_type.name.lower(), name)
-        recomp_addr = self._find_potential_match(name, compare_type)
-        if recomp_addr is None:
-            return False
+        for obj in self.search_name(name, compare_type):
+            if obj.orig_addr is None and obj.recomp_addr is not None:
+                matched = self.set_pair(addr, obj.recomp_addr, compare_type)
+
+                # Type field has been set by set_pair, so we can use it in our count query:
+                (count,) = self._sql.execute(
+                    """SELECT count(rowid) from symbols
+                    where json_extract(kvstore,'$.name') = ?
+                    AND json_extract(kvstore,'$.type') = ?""",
+                    (name, compare_type),
+                ).fetchone()
 
-        return self.set_pair(addr, recomp_addr, compare_type)
+                if matched and count > 1:
+                    logger.warning(
+                        "Ambiguous match 0x%x on name '%s' to '%s'",
```

Comment:
What do you think about logging *all* matches for an ambiguous name instead of logging just the one that is chosen automatically? It might help finding the correct one faster, though you'd probably have to look into the cvdump in any case.

Reply:
The way it is now, it will warn for each match if the name and type appear more than once. It will warn even if we pick the only unmatched option remaining. We don't warn if the match failed (because the orig. address was already used.)

Do you mean displaying a summary of each non-unique name/type combination at the end? For example: if there are 5 `_Construct` functions and we matched 3 of them, we may have picked incorrectly 3 times.

Reply:
Let's say we have one annotation with an ambiguous name `examplefn` (maybe a `// LIBRARY`), and there are two symbols with that name (lets say `?SYMBOL1` and `?SYMBOL2`). The way I understand the current code, it would output one warning `Ambiguous match 0xabcdabcd on name 'examplefun' to '?SYMBOL1`, but `?SYMBOL2` never makes it to the logger. I was wondering if it makes sense to log `Ambiguous match 0xabcdabcd on name 'examplefun' to '?SYMBOL1, could also refer to ?SYMBOL2, ...`.

The current code looks very reasonable if the number of annotations matches the number of symbols; my idea might improve the developer experience in other situations.

Reply:
Yeah, that's a good idea. I also think we should _not_ warn if there are multiple options but you have eliminated all but one by matching on either the line number or a symbol (i.e. match types without ambiguity). This doesn't happen now because the code parser doesn't properly capture the function's name, but that will change with the lexer/token-based parser.

This would require changing the approach to display a summary at the end for each duplicate name/type in the DB if:
- At least one was matched by name and any are unmatched
- More than one matched by name

I think that's easier than it sounds so I could do it now, or we can push this as a stopgap to fix the outstanding problems. What do you think?

Reply:
I like your idea, but I think it makes more sense to implement it when the parser refactor has been completed.